Gordon360/Services/LostAndFoundService.cs

```diff
@@ -0,0 +1,338 @@
+            {
+                report = context.MissingItemData.FirstOrDefault(x => x.ID == id);
+            }
+            else
+            {
+                // Otherwise check if the report belongs to the requesting user
+                var idNum = context.ACCOUNT.FirstOrDefault(x => x.AD_Username == username).gordon_id;
+                report = context.MissingItemData.FirstOrDefault(x => x.ID == id && x.submitterID == idNum);
+            }
+            // Typecast the data into the viewmodel
+            MissingItemReportViewModel returnReport = (MissingItemReportViewModel)report;
+
+            // Get the list of public admin actions on this report, and add them to the report.
+            returnReport.adminActions = GetActionsTaken(id, true);
+            
+            return returnReport;
+        }
+
+        /// <summary>
+        /// Gets a list of Actions Taken by id
+        /// </summary>
+        /// <param name="id">The ID of the associated missing item report</param>
+        /// <param name="getPublicOnly">Oonly get actions marked as public.  Default false.</param>
+        /// <returns>An ActionsTaken, or null if no item matches the id</returns>
+        public IEnumerable<ActionsTakenViewModel> GetActionsTaken(int id, bool getPublicOnly = false)
+        {
+            IEnumerable<ActionsTaken> actionsList;
+            if (getPublicOnly)
+            {
+                actionsList = context.ActionsTaken.Where(x => x.missingID == id && x.isPublic);
+            }
+            else
+            {
+                actionsList = context.ActionsTaken.Where(x => x.missingID == id);
+            }
+
+            // Create a list of usernames based on the submitter ID
+            List<string> usernameList = [];
+            foreach (ActionsTaken element in actionsList)
+            {
+                var submitter_id = context.ACCOUNT.FirstOrDefault(x => x.gordon_id == element.submitterID);
+
+                string username;
+
+                if (submitter_id != null)
+                {
+                    username = submitter_id.AD_Username;
+                    usernameList.Add(username);
+                }
+                else
+                {
+                    throw new ResourceCreationException() { ExceptionMessage = "No username could be found for the user." };
+                }
+            }
+
+            // Type cast to Actions Taken View Model 
+            IEnumerable<ActionsTakenViewModel> returnList = actionsList.Select(x => (ActionsTakenViewModel)x);
+
+            // Create an Enumerator for the username list
+            IEnumerator<string> usernameEnumerator = usernameList.GetEnumerator();
+
+            // For each action taken in returnList, update the username based on the username list enumerator
+            returnList = returnList.Select(x => { usernameEnumerator.MoveNext(); x.username = usernameEnumerator.Current; return x; });
+
+            return returnList;
+        }
```

Comment:
You should replace this messy username logic with a view of `actionsTaken` that has usernames.

The view could look like:
```sql
select 
  ActionsTaken.* -- (don't use *, write out the actual columns from ActionsTaken that you need)
  , Account.AD_Username
from CCT.LostAndFound.ActionsTaken
LEFT JOIN CCT.dbo.Account on ActionsTaken.submitterId = Account.gordon_id
```

Reply:
You can also do this with a Join in C#, so that you don't have to manage the additional SQL object, if you prefer:
```C#
// LostAndFoundService.GetActionsTaken:
            IQueryable<ActionsTaken> actionsList = context.ActionsTaken.AsQueryable();

            if (getPublicOnly)
            {
                actionsList = actionsList.Where(x => x.isPublic);
            }

            return actionsList.Join(context.ACCOUNT, at => at.submitterID, acc => acc.gordon_id, (at, acc) => ActionsTakenViewModel.From(at, acc.AD_Username));

// ActionsTakenViewModel.From:
        public static ActionsTakenViewModel From(CCT.ActionsTaken ActionTaken, string SubmitterUsername)
            => new ActionsTakenViewModel
            {
                ID = ActionTaken.ID,
                missingID = ActionTaken.missingID,
                action = ActionTaken.action,
                actionDate = ActionTaken.actionDate,
                actionNote = ActionTaken.actionNote,
                isPublic = ActionTaken.isPublic,
                username = SubmitterUsername
            };
```


That results in the same SQL query:
```sql
SELECT [a].[ID], [a].[action], [a].[actionDate], [a].[actionNote], [a].[isPublic], [a].[missingID], [a].[submitterID], [a0].[AD_Username]
FROM [LostAndFound].[ActionsTaken] AS [a]
INNER JOIN [dbo].[ACCOUNT] AS [a0] ON [a].[submitterID] = [a0].[gordon_id]
```

Reply:
Cool, I wrote a SQL view to do add the usernames just like you said, that's how we implemented it for the missing item report, but I think we just didn't think about it for the actions taken.

Reply:
Added view in API #1111